### PR TITLE
Fix blocking sends and message boundary issues

### PR DESCRIPTION
**Pull request by asahilina**
This is kind of subtle. We need to correctly handle:

- Packet tails for fully parsed packet headers (these can be sent incrementally, but importantly should *not* be merged with any fd-carrying packets because libxcb cannot handle that, as fds are logically attached to the first byte sent)
- Packet "heads" which are partial unparsed packet headers, which we defer parsing to when we get more data. These might have fds attached, which we also need to stash for later.

This introduces a bunch of corner cases, so this commit attempts to be more defensive and introduces a bunch of asserts.

And we also need to not block on sendmsg (which can deadlock), so to fix that push sends into a queue that is processed as part of the epoll loop. This is probably easier than trying to rewrite the whole app in async rust. Also fix partial sendmsg calls by re-queueing the remainder if needed.

For reference, the libxcb behavior is that it will receive any number of fds when receiving packet headers, and it will only consume as many as needed for each processed reply, but it will *not* receive fds (it uses plain recv()) when processing its version of packet tails. So that means we always need to send fds when the first byte of the send buffer is the first byte of a packet, but it is okay to send "too many" and later send more packets that consume them (this makes the partial sendmsg behavior correct).

The way this is implemented, all this complex logic goes in x112virtgpu (for both directions) and we don't touch the rutabaga side.

Reference for the subtleties of fd passing and receive/send splitting: https://unix.stackexchange.com/questions/185011/what-happens-with-unix-stream-ancillary-data-on-partial-reads

With these fixes and an unrelated fix to the libkrun side for host->guest dmabuf passing, Steam finally runs stable and Portal 2 works.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1346,78 +1487,120 @@ fn main() {
...
             }
         }
-        let fd = evts[0].data();
-        if fd == listen_sock.as_raw_fd() as u64 {
-            let res = listen_sock.accept();
-            if res.is_err() {
-                eprintln!(
-                    "Failed to accept a connection, error: {:?}",
-                    res.unwrap_err()
+        for evt in evts {
```
Comment (WhatAmISupposedToPutHere) [inline src/main.rs:1490]: not sure about the point of this change, only makes it harder to review

Comment (asahilina) [reply src/main.rs:1490]: Should just be more efficient than looping again and going through epoll for no reason.

Comment (WhatAmISupposedToPutHere) [reply src/main.rs:1490]: But `evts` always has 1 element? If you want to grab multiple, you need to put multiple event slots there.

Comment (asahilina) [reply src/main.rs:1490]: Oh, I missed that...

Comment (asahilina) [reply src/main.rs:1490]: Fixed ^^

Comment (WhatAmISupposedToPutHere) [reply src/main.rs:1490]: You also need to get the return value from epoll.wait, and if it is `Ok(i)`, you process only the first `i` elements of the evts array. (the rest are left as-is and are invalid)

Comment (asahilina) [reply src/main.rs]: I didn't think it mattered since the unwritten events would not match any of the branches, but yeah, makes sense.

**Pull request merged**